[usage] Persist usage reports in database

andrew-farries opened the pull request:
## Description

Store the usage report in the database table created in #11311 whenever usage reconciliation runs.

⚠️ This PR does not yet populate the `creditsUsed` or `generationId` fields ⚠️ 

## Related Issue(s)

Part of #10323 

## How to test

Run the usage component against the database in this PR's preview environment:

* Port forward to the database in the preview env
* Edit the local config `components/usage/config.json` for the usage controller to run every 10s.
* Run the usage component:
```bash
DB_USERNAME=gitpod DB_PASSWORD=<> DB_HOST=localhost DB_PORT=3306 go run . run
```
(get the password from `server` environment)
* Wait a while for reconciliation to run a couple of times.
* Connect to the port-forwarded database with eg `mycli` and observe the data in `d_b_workspace_instance_usage`.

Unit tests.

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-preview

easyCZ commented on components/usage/pkg/db/workspace_instance_usage_test.go:
```diff
@@ -0,0 +1,102 @@
+							// the same workspace again
+							ID:                 instanceID,
+							UsageAttributionID: teamAttributionID,
+							WorkspaceClass:     "default",
```
There should be a constant for this `default` value

easyCZ commented on components/usage/pkg/db/workspace_instance_usage_test.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+func TestCanHandleMultipleReportsWithDuplicateEntries(t *testing.T) {
+	conn := dbtest.ConnectForTests(t)
```
I'd recommend moving the `conn` into each test scenario, to ensure that after each test the data is deleted before the next test run

andrew-farries replied:
done

easyCZ commented on components/usage/pkg/controller/reconciler.go:
```diff
@@ -85,6 +85,8 @@ func (u *UsageReconciler) Reconcile() (err error) {
 	}
 	log.Infof("Wrote usage report into %s", filepath.Join(dir, stat.Name()))
 
+	db.CreateUsageRecords(ctx, u.conn, report)
```
Should handle error

easyCZ commented on components/usage/pkg/db/workspace_instance_usage.go:
```diff
@@ -0,0 +1,61 @@
+	return "d_b_workspace_instance_usage"
+}
+
+func CreateUsageRecords(ctx context.Context, conn *gorm.DB, report map[AttributionID][]WorkspaceInstanceForUsage) error {
```
Or actually, it should take `[]WorkspaceInstanceUsage`

andrew-farries replied:
done. The conversion between reports and `[]WorkspaceInstanceUsage` is done in the reconciler.

andrew-farries commented on components/usage/pkg/db/workspace_instance_usage_test.go:
```diff
@@ -0,0 +1,102 @@
+	for _, scenario := range scenarios {
+		t.Run(scenario.Name, func(t *testing.T) {
+			for _, report := range scenario.Reports {
+				require.NoError(t, db.CreateUsageRecords(ctx, conn, report))
```
I've added `t.Cleanup` calls to each subtest to delete the usage records between each subtest.

easyCZ replied:
Commented in https://github.com/gitpod-io/gitpod/pull/11343/files#r919996694

easyCZ commented on components/usage/pkg/db/workspace_instance_usage_test.go:
```diff
@@ -0,0 +1,112 @@
+		conn := dbtest.ConnectForTests(t)
+		t.Run(scenario.Name, func(t *testing.T) {
+			t.Cleanup(func() {
+				require.NoError(t, conn.Where("1=1").Delete(&db.WorkspaceInstanceUsage{}).Error)
```
This was the cause of the problems with flaky tests. When more tests exist and do this, it deletes the data needed for other runs of the test. That's why the current approach adds a hook for each ID created to only deleted the IDs created in each test.

andrew-farries replied:
🤔 

None of these tests call `t.Parallel`, so AIUI the tests in this file will run in series, and within each test each subtest will also run in series. So where is the risk of having a subtest clearing the table after it finishes?

Is it tests running in different packages (which will be run in parallel by `go test`) that is the concern here?

easyCZ replied:
https://github.com/gitpod-io/gitpod/pull/10971 has extra context on this, including a link to some literature

andrew-farries replied:
> But tests in different files will be run concurrently.

I don't think that's correct. See for example [here](https://engineering.mercari.com/en/blog/entry/20220408-how_to_use_t_parallel/):

> By default, execution of test code using the testing package will be done sequentially. However, note that it is only the tests within a given package that run sequentially.

> If tests from multiple packages are specified, the tests will be run in parallel at the package level. For example, imagine there are two packages, package a and package b. The test code in package a will be run sequentially, and the test code in package b will also be run sequentially. However, the tests for package a and package b will be run in parallel. Let’s look more closely into how these will be run in parallel.

So I guess I'm struggling to understand how test cleanup code within a package can interfere with other tests in the same package if there is no `t.Parallel` anywhere. Maybe there is a race in actually deleting the rows from the database?

easyCZ replied:
The issues we observed were across different packages. And the same will happen here once we add proper tests to the controller package.

andrew-farries replied:
test cleanup changed to only remove records that were created by each test.

geropl commented on components/usage/pkg/db/workspace_instance_usage.go:
```diff
@@ -0,0 +1,39 @@
+)
+
+type WorkspaceInstanceUsage struct {
+	WorkspaceID   uuid.UUID     `gorm:"primary_key;column:workspaceId;type:char;size:36;" json:"workspaceId"`
```
Does it make sense to merge the other PR with the rename into this one...? I though it had already been merged.

:bulb: Or has the table defintion, but not the Go struct?

geropl commented:
Doing a quick test...

geropl commented:
Works as advertised. I understood that credits and generationId will be set later.

Only thing I noticed is that `stoppedAt` is not set, while the workspace_instance is definitely `stopped` (and has a `stoppedTime`).
![image](https://user-images.githubusercontent.com/32448529/178926065-a553d543-269d-43e3-b68b-f93120c86f91.png)

@andrew-farries Any idea why that might be?

andrew-farries commented:
> Only thing I noticed is that stoppedAt is not set, while the workspace_instance is definitely stopped (and has a stoppedTime).

Thanks. I've fixed this and added tests for the conversion from a usage report to instance usage records. (f25999dcf and b9016e9de).

geropl reviewed (approved):
Note that the current preview env is in [an undefined state](https://werft.gitpod-dev.com/job/gitpod-build-af-store-usage-data.11).

But I applied the migration manually, ran the test and :tada: : it worked. :lotus_position: 
![image](https://user-images.githubusercontent.com/32448529/178975921-d24bf56a-12d1-434b-aa91-7472b324e449.png)

andrew-farries commented:
/werft run with-clean-slate-deployment=true with-preview=true

liam-j-bennett commented:
/hold this is blocking the merge queue of a fix that fixes the merge queue 😢 I'll unhold when I've merged

liam-j-bennett commented:
/unhold You will need to restart the werft job. Apologies for the downtime 🤦

geropl commented:
@andrew-farries What about runnign `with-preview=false`? I already tested twice and am confident it works, so let's get this in! :ship:

andrew-farries commented:
/unhold
